internal/common/auth/authorization/kubernetes.go

```diff
@@ -0,0 +1,206 @@
+package authorization
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware/util/metautils"
+	"github.com/patrickmn/go-cache"
+	log "github.com/sirupsen/logrus"
+	authv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/G-Research/armada/internal/common/auth/configuration"
+)
+
+type KubernetesNativeAuthService struct {
+	KidMappingFileLocation string
+	TokenCache             *cache.Cache
+	InvalidTokenExpiry     int64
+}
+
+func NewKubernetesNativeAuthService(config configuration.KubernetesAuthConfig) KubernetesNativeAuthService {
+	cache := cache.New(5*time.Minute, 5*time.Minute)
+	return KubernetesNativeAuthService{
+		KidMappingFileLocation: config.KidMappingFileLocation,
+		TokenCache:             cache,
+		InvalidTokenExpiry:     config.InvalidTokenExpiry,
+	}
+}
+
+type CacheData struct {
+	Name  string `json:"name"`
+	Valid bool   `json:"valid"`
+}
+
+func (authService *KubernetesNativeAuthService) Authenticate(ctx context.Context) (Principal, error) {
+	// Retrieve token from context.
+	authHeader := strings.SplitN(metautils.ExtractIncoming(ctx).Get("authorization"), " ", 2)
+
+	if len(authHeader) < 2 || authHeader[0] != "KubernetesAuth" {
+		return nil, missingCredentials
+	}
+
+	token, ca, err := parseAuth(authHeader[1])
+	if err != nil {
+		return nil, missingCredentials
+	}
+
+	// Get token time
+	expirationTime, err := parseTime(token)
+	if err != nil {
+		return nil, err
+	}
+
+	if time.Now().After(*expirationTime) {
+		return nil, fmt.Errorf("invalid token, expired")
+	}
+
+	// Check Cache
+	data, found := authService.TokenCache.Get(token)
+	if found {
+		if cacheInfo, ok := data.(CacheData); ok {
+			if cacheInfo.Valid {
+				return NewStaticPrincipal(cacheInfo.Name, []string{cacheInfo.Name}), nil
+			} else {
+				return nil, fmt.Errorf("token invalid")
+			}
+		}
+	}
+
+	// Get URL from token KID
+	url, err := authService.getClusterURL(token)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make request to token review endpoint
+	name, err := authService.reviewToken(ctx, url, token, []byte(ca))
+	if err != nil {
+		return nil, err
+	}
+
+	// Add to cache
+	authService.TokenCache.Set(
+		token,
+		CacheData{
+			Name:  name,
+			Valid: true,
+		},
+		expirationTime.Sub(time.Now()))
+
+	// Return very basic Principal
+	return NewStaticPrincipal(name, []string{name}), nil
+}
+
+func (authService *KubernetesNativeAuthService) getClusterURL(token string) (string, error) {
+	log.Infof("Token: %s", token)
```

Comment:
Should we log these things?  This seems like a security vulnerability?

Reply:
We indeed should not be logging these things, am in the process of taking them out